File: apps/sensor-display-as/assembly/sht4x.ts
// SHT40（SHT4x）の読み出し。apps/README.md §1 が正。
//
// sensor-display-rs の sht4x.rs と同じ host call 列・同じ値を出す。
// 片方だけ変えると sensor_display_rs_and_as_agree が落ちる。

import { I2cBus, time } from "../../../bindings/assemblyscript/assembly/index";

/// I2C アドレス。
export const ADDRESS: u16 = 0x44;

/// 計測が終わるまでの待ち時間。データシートの高精度計測は最大 8.3 ms。
const MEASURE_MS: u32 = 10;

/// CRC-8。多項式 0x31、初期値 0xFF、反転なし。
export function crc8(data: Uint8Array, from: i32, len: i32): u8 {
  let crc: u8 = 0xff;
  for (let i = 0; i < len; i++) {
    crc ^= data[from + i];
    for (let b = 0; b < 8; b++) {
      crc = (crc & 0x80) != 0 ? <u8>((crc << 1) ^ 0x31) : <u8>(crc << 1);
    }
  }
  return crc;
}

/// 読み出しの結果。`ok` が false のときは `crcFailed` で理由を分ける。
export class Reading {
  ok: bool = false;
  crcFailed: bool = false;
  rawT: u16 = 0;
  rawH: u16 = 0;
}

const MEASURE = new Uint8Array(1);
const FRAME = new Uint8Array(6);

/// 1 回測って読む。
export function read(bus: I2cBus): Reading {
  const r = new Reading();
  // SHT4x は 1 バイトコマンド（高精度計測）。
  MEASURE[0] = 0xfd;
  if (bus.write(ADDRESS, MEASURE) != 0) return r;
  time.sleepMs(MEASURE_MS);

  if (bus.read(ADDRESS, FRAME) != 6) return r;
  if (crc8(FRAME, 0, 2) != FRAME[2] || crc8(FRAME, 3, 2) != FRAME[5]) {
    r.crcFailed = true;
    return r;
  }
  r.ok = true;
  r.rawT = <u16>((<u16>FRAME[0] << 8) | <u16>FRAME[1]);
  r.rawH = <u16>((<u16>FRAME[3] << 8) | <u16>FRAME[4]);
  return r;
}

/// 温度（摂氏 ×100）。固定小数で計算する。
export function tempCenti(rawT: u16): i32 {
  return -4500 + (17500 * <i32>rawT) / 65535;
}

/// 相対湿度（％ ×100）。
///
/// SHT4x の式は raw=0 で -600、raw=65535 で 11900 を返すので、データシートの
/// とおり 0..10000 に収める。min / max を使わず if で書くのは Rust 版と
/// 同じ命令列にするため（apps/README.md §1）。
export function humidityCenti(rawH: u16): i32 {
  let v = -600 + (12500 * <i32>rawH) / 65535;
  if (v < 0) v = 0;
  if (v > 10000) v = 10000;
  return v;
}

File: apps/sensor-display-rs/src/sht4x.rs
//! SHT40（SHT4x）の読み出し。`apps/README.md` §1 が正。
//!
//! `sensor-display-as` の `sht4x.ts` と**同じ host call 列・同じ値**を出す。
//! 片方だけ変えると `sensor_display_rs_and_as_agree` が落ちる。

use wasmicon_hal::i2c::Bus;
use wasmicon_hal::time;

/// I2C アドレス。
pub const ADDRESS: u16 = 0x44;

/// 単発計測（高精度）。SHT4x は 1 バイトコマンド。
const MEASURE: [u8; 1] = [0xFD];

/// 計測が終わるまでの待ち時間。データシートの高精度計測は最大 8.3 ms。
const MEASURE_MS: u32 = 10;

/// CRC-8。多項式 0x31、初期値 0xFF、反転なし。
#[must_use]
pub fn crc8(data: &[u8]) -> u8 {
    let mut crc: u8 = 0xFF;
    for &b in data {
        crc ^= b;
        let mut i = 0;
        while i < 8 {
            crc = if crc & 0x80 != 0 {
                (crc << 1) ^ 0x31
            } else {
                crc << 1
            };
            i += 1;
        }
    }
    crc
}

/// 生の測定値。
pub struct Reading {
    pub raw_t: u16,
    pub raw_h: u16,
}

/// 読み出しの失敗。
pub enum Error {
    /// 通信できない。
    Io,
    /// CRC が合わない。
    Crc,
}

/// 1 回測って読む。
///
/// # Errors
/// 通信に失敗したら `Io`、CRC が合わなければ `Crc`。
pub fn read(bus: &Bus) -> Result<Reading, Error> {
    if bus.write(ADDRESS, &MEASURE).is_err() {
        return Err(Error::Io);
    }
    time::sleep_ms(MEASURE_MS);

    let mut buf = [0u8; 6];
    match bus.read(ADDRESS, &mut buf) {
        Ok(6) => {}
        _ => return Err(Error::Io),
    }
    if crc8(&buf[0..2]) != buf[2] || crc8(&buf[3..5]) != buf[5] {
        return Err(Error::Crc);
    }
    Ok(Reading {
        raw_t: (u16::from(buf[0]) << 8) | u16::from(buf[1]),
        raw_h: (u16::from(buf[3]) << 8) | u16::from(buf[4]),
    })
}

/// 温度（摂氏 ×100）。固定小数で計算する。
#[must_use]
pub fn temp_centi(raw_t: u16) -> i32 {
    -4500 + (17500 * i32::from(raw_t)) / 65535
}

/// 相対湿度（％ ×100）。
///
/// SHT4x の式は raw=0 で -600、raw=65535 で 11900 を返すので、データシートの
/// とおり 0..10000 に収める。`clamp` や `min` / `max` を使わず `if` で書くのは
/// AssemblyScript 版と同じ命令列にするため（`apps/README.md` §1）。
/// `lib.rs` の `bar_px` と同じ理由で clippy の勧めには従わない。
#[allow(clippy::manual_clamp)]
#[must_use]
pub fn humidity_centi(raw_h: u16) -> i32 {
    let mut v = -600 + (12500 * i32::from(raw_h)) / 65535;
    if v < 0 {
        v = 0;
    }
    if v > 10000 {
        v = 10000;
    }
    v
}
